sycl/source/detail/graph_impl.cpp

```diff
@@ -226,6 +228,133 @@ bool graph_impl::clearQueues() {
+/// will return.
+/// @return True if the search should end immediately, false if not.
+bool visitNodeDepthFirst(
+    std::shared_ptr<node_impl> Node,
+    std::set<std::shared_ptr<node_impl>> &VisitedNodes,
+    std::deque<std::shared_ptr<node_impl>> &NodeStack,
+    std::function<bool(std::shared_ptr<node_impl> &,
+                       std::deque<std::shared_ptr<node_impl>> &)>
+        NodeFunc) {
+  auto EarlyReturn = NodeFunc(Node, NodeStack);
+  if (EarlyReturn) {
+    return true;
+  }
+  NodeStack.push_back(Node);
+  Node->MVisited = true;
+  VisitedNodes.emplace(Node);
+  for (auto &Successor : Node->MSuccessors) {
+    if (visitNodeDepthFirst(Successor, VisitedNodes, NodeStack, NodeFunc)) {
+      return true;
+    }
+  }
+  NodeStack.pop_back();
+  return false;
+}
+
+void graph_impl::searchDepthFirst(
+    std::function<bool(std::shared_ptr<node_impl> &,
+                       std::deque<std::shared_ptr<node_impl>> &)>
+        NodeFunc) {
+  std::set<std::shared_ptr<node_impl>> VisitedNodes;
+
+  for (auto &Root : MRoots) {
+    std::deque<std::shared_ptr<node_impl>> NodeStack;
+    if (visitNodeDepthFirst(Root, VisitedNodes, NodeStack, NodeFunc)) {
+      break;
+    }
+  }
+
+  // Reset the visited status of all nodes encountered in the search.
+  for (auto &Node : VisitedNodes) {
+    Node->MVisited = false;
+  }
+}
+
+bool graph_impl::checkForCycles() {
+  // Using a depth-first search and checking if we vist a node more than once in
+  // the current path to identify if there are cycles.
+  bool CycleFound = false;
+  auto CheckFunc = [&](std::shared_ptr<node_impl> &Node,
+                       std::deque<std::shared_ptr<node_impl>> &NodeStack) {
+    // If the current node has previously been found in the current path through
+    // the graph then we have a cycle and we end the search early.
+    if (std::find(NodeStack.begin(), NodeStack.end(), Node) !=
+        NodeStack.end()) {
+      CycleFound = true;
+      return true;
+    }
+    return false;
+  };
+  searchDepthFirst(CheckFunc);
+  return CycleFound;
+}
+
+void graph_impl::makeEdge(std::shared_ptr<node_impl> Src,
+                          std::shared_ptr<node_impl> Dest) {
+  if (MRecordingQueues.size()) {
+    throw sycl::exception(make_error_code(sycl::errc::invalid),
+                          "make_edge() cannot be called when a queue is "
+                          "currently recording commands to a graph.");
+  }
+  if (Src == Dest) {
+    throw sycl::exception(
+        make_error_code(sycl::errc::invalid),
+        "make_edge() cannot be called when Src and Dest are the same.");
+  }
+
+  bool SrcFound = false;
+  bool DestFound = false;
+  auto CheckForNodes = [&](std::shared_ptr<node_impl> &Node,
+                           std::deque<std::shared_ptr<node_impl>> &) {
+    if (Node == Src) {
+      SrcFound = true;
+    }
+    if (Node == Dest) {
+      DestFound = true;
+    }
+    return SrcFound && DestFound;
+  };
```

Comment:
If we want to keep the undefined behaviour for the "finalize" step as mentioned in the specification. We could check if Node is already present in "NodeStack" (second param) to break the recursive search if the graph includes a cycle.

Reply:
I'm not sure I understand what you mean, do you mean we could combine the searches and just check for both presence of nodes and cycles? That seems possible to me. 

If that's what you mean though I don't understand what that has to do with preserving the graph with cycles until `finalize()`. To do that we would still need to modify the behaviour for these cases specifically which doesn't seem worthwhile to me.

Reply:
In fact, I made this comment before discussing on Zulip the relevance of offering the no-cycle-check feature.
With the current implementation a cycle in the graph could result in an "infinite" recursive loop.
Just wanted to point out that this risk could be removed by adding an additional check and break condition using NodeStack (similar to the one you did to check cycles).

Reply:
OK I think I understand, if the user has disabled cycle checks and introduces a cycle here then the check could potentially end up in an infinite loop. However, I'm reluctant to add error handling for this case since this is undefined behaviour, and would make the search for nodes potentially as expensive or more than the check for cycles with no way to remove that behaviour. Since it is only needed when we're in undefined territory I don't think the tradeoff is worth it there.

Reply:
Makes sense